### ISSUE TITLE
Feature request: some optimization on pagination

### ISSUE THREAD

**Issue by razbenya** [arangodb/arangodb#13476]
## My Environment

* __ArangoDB Version__:         3.7.6

## Component, Query & Data

__AQL query (if applicable)__:
```
for record in v_records
    SEARCH ANALYZER(record.queries in @queries , "text_en")
    SORT record.score desc
    SORT @start,1000
    return record._key
```
__AQL explain and/or profile (if applicable)__:

profile for first page 
```
Query String (149 chars, cacheable: false):
 for record in v_records
     SEARCH ANALYZER(record.queries in @queries, "text_en")
     sort record.score desc
     limit 0,1000
     return record._key

Execution plan:
 Id   NodeType            Calls   Items   Runtime [s]   Comment
  1   SingletonNode           1       1       0.00001   * ROOT
  2   EnumerateViewNode       1    1000       0.00528     - FOR record IN v_records SEARCH ANALYZER((record.`queries` == "a"), "text_en") SORT record.`score` DESC   /* view query */
  5   LimitNode               1    1000       0.00002       - LIMIT 0, 1000
  6   CalculationNode         1    1000       0.00028       - LET #3 = record.`_key`   /* attribute expression */
  7   ReturnNode              1    1000       0.00001       - RETURN #3

Indexes used:
 none

Optimization rules applied:
 Id   RuleName
  1   handle-arangosearch-views
  2   remove-unnecessary-calculations-2

Query Statistics:
 Writes Exec   Writes Ign   Scan Full   Scan Index   Filtered   Exec Time [s]
           0            0           0         1000          0         0.00616

Query Profile:
 Query Stage           Duration [s]
 initializing               0.00000
 parsing                    0.00014
 optimizing ast             0.00001
 loading collections        0.00002
 instantiating plan         0.00004
 optimizing plan            0.00024
 executing                  0.00562
 finalizing                 0.00009
```
profile for page with 9m skip
```
Query String (155 chars, cacheable: false):
 for record in v_records
     SEARCH ANALYZER(record.queries in @queries, "text_en")
     sort record.score desc
     limit 9000000,1000
     return record._key

Execution plan:
 Id   NodeType            Calls     Items   Runtime [s]   Comment
  1   SingletonNode           1         1       0.00001   * ROOT
  2   EnumerateViewNode       1   9001000       0.44689     - FOR record IN v_records SEARCH ANALYZER((record.`queries` == "a"), "text_en") SORT record.`score` DESC   /* view query */
  5   LimitNode               1      1000       0.00002       - LIMIT 9000000, 1000
  6   CalculationNode         1      1000       0.00028       - LET #3 = record.`_key`   /* attribute expression */
  7   ReturnNode              1      1000       0.00001       - RETURN #3

Indexes used:
 none

Optimization rules applied:
 Id   RuleName
  1   handle-arangosearch-views
  2   remove-unnecessary-calculations-2

Query Statistics:
 Writes Exec   Writes Ign   Scan Full   Scan Index   Filtered   Exec Time [s]
           0            0           0      9001000          0         0.44765

Query Profile:
 Query Stage           Duration [s]
 initializing               0.00000
 parsing                    0.00008
 optimizing ast             0.00001
 loading collections        0.00002
 instantiating plan         0.00003
 optimizing plan            0.00015
 executing                  0.44723
 finalizing                 0.00012
```
profile after 9001000 skips

```
Query String (155 chars, cacheable: false):
 for record in v_records
     SEARCH ANALYZER(record.queries in @queries, "text_en")
     sort record.score desc
     limit 9001000,1000
     return record._key

Execution plan:
 Id   NodeType            Calls     Items   Runtime [s]   Comment
  1   SingletonNode           1         1       0.00001   * ROOT
  2   EnumerateViewNode       1   9002000       0.40489     - FOR record IN v_records SEARCH ANALYZER((record.`queries` == "a"), "text_en") SORT record.`score` DESC   /* view query */
  5   LimitNode               1      1000       0.00002       - LIMIT 9001000, 1000
  6   CalculationNode         1      1000       0.00027       - LET #3 = record.`_key`   /* attribute expression */
  7   ReturnNode              1      1000       0.00001       - RETURN #3

Indexes used:
 none

Optimization rules applied:
 Id   RuleName
  1   handle-arangosearch-views
  2   remove-unnecessary-calculations-2

Query Statistics:
 Writes Exec   Writes Ign   Scan Full   Scan Index   Filtered   Exec Time [s]
           0            0           0      9002000          0         0.40552

Query Profile:
 Query Stage           Duration [s]
 initializing               0.00000
 parsing                    0.00006
 optimizing ast             0.00001
 loading collections        0.00001
 instantiating plan         0.00002
 optimizing plan            0.00011
 executing                  0.40521
 finalizing                 0.00008
```

__Dataset__:

records has 10m entities after the SEARCH operation, and score field is the view natural order

__Problem__:
Getting the first 1k entities is very fast, than its getting slower as I proceed further with the paging, 
As I can see in the query profile all entities are getting scanned when using skip. 

I think this can be optimized if arango somehow remember the last page,  Is something like this possible?

**Comment by razbenya** [comment]
Also I tried to use CURSOR API instead and do my paging to cursor, seems to take really long time until the cursor is ready, 
the expectation was that the cursor will be returned as soon as the first batch is available but it is not the case.

**Comment by goedderz** [comment]
Hi @razbenya,

> the expectation was that the cursor will be returned as soon as the first batch is available

a streaming cursor does just that. See the `stream` option in the cursor API.

Best,

Tobias

**Comment by razbenya** [comment]
Hi Tobias thanks for the answer,

I tried the stream option and it was really slow it took more than twice as without the option or iterating with offset and limit over all results, 
also the documentation recommended not using the stream options for long queries.

Raz.

**Comment by goedderz** [comment]
> I tried the stream option and it was really slow it took more than twice as without the option or iterating with offset and limit over all results,

That is unexpected. Can you provide more details what you tried exactly, and the timing measurements you did? I.e. the queries you executed, the client code you executed them with, query profiles etc.

> also the documentation recommended not using the stream options for long queries.

That is an outdated comment and was targeted at the MMFiles storage engine, which is no longer supported. Unless you're using exclusive locks or so, you can safely ignore it and use streaming queries also for long running queries.

Best,

Tobias

**Comment by razbenya** [comment]
The data set and the query is similar to what I present in my original post: 
I added Limit to iterate over max 2.5M entties: 

cursor with stream took around 40 seconds to iterate over all cursor. 
cursor without stream:  took 12 seconds till cursor was available than another 2 seconds to iterate  (14 seconds overall)

I am using java driver version 6.8.2
test was against single arango server version 3.7.6 installed locally.

**Comment by goedderz** [comment]
Can you share profiles of the query, once with and once without streaming cursor? You can do that with arangosh, e.g. by calling `db._profileQuery`. The third parameter are options, there you can set `stream` to either `true` or `false` (or omit it).

**Comment by razbenya** [comment]
stream: false: 
```
Query String (117 chars, cacheable: false):
 for record in v_records
     SEARCH ANALYZER(record.queries in ["a"], "text_en")
     LIMIT 2500000
     return record._key

Execution plan:
 Id   NodeType            Calls     Items   Runtime [s]   Comment
  1   SingletonNode           1         1       0.00001   * ROOT
  2   EnumerateViewNode    2500   2500000       9.95593     - FOR record IN v_records SEARCH ANALYZER((record.`queries` == "a"), "text_en")   /* view query */
  3   LimitNode            2500   2500000       0.03155       - LIMIT 0, 2500000
  4   CalculationNode      2500   2500000       0.62575       - LET #1 = record.`_key`   /* attribute expression */
  5   ReturnNode           2500   2500000       0.07697       - RETURN #1

Indexes used:
 none

Optimization rules applied:
 Id   RuleName
  1   handle-arangosearch-views

Query Statistics:
 Writes Exec   Writes Ign   Scan Full   Scan Index   Filtered   Exec Time [s]
           0            0           0      2500000          0        10.69257

Query Profile:
 Query Stage           Duration [s]
 initializing               0.00000
 parsing                    0.00013
 optimizing ast             0.00001
 loading collections        0.00002
 instantiating plan         0.00004
 optimizing plan            0.00019
 executing                 10.69199
 finalizing                 0.00017
```

when I am trying the same thing with stream true I am getting error 

```
127.0.0.1:8529@test> db._profileQuery(` for record in v_records     SEARCH ANALYZER(record.queries in ["a"], "text_en")     LIMIT 2500000 return record._key`,  {},  {"stream": true} );
JavaScript exception in file '/usr/share/arangodb3/js/common/modules/@arangodb/aql/explainer.js' at 766,22: TypeError: Cannot read property 'nodes' of undefined
!  recursiveWalk(plan.nodes, 0, 'COOR');
!                     ^
stacktrace: TypeError: Cannot read property 'nodes' of undefined
    at processQuery (/usr/share/arangodb3/js/common/modules/@arangodb/aql/explainer.js:766:22)
    at Object.profileQuery (/usr/share/arangodb3/js/common/modules/@arangodb/aql/explainer.js:2076:3)
    at Proxy.ArangoDatabase._profileQuery (/usr/share/arangodb3/js/client/modules/@arangodb/arango-database.js:968:38)
    at <shell command>:1:4
```

**Comment by goedderz** [comment]
Thanks, and sorry for that. There seems to be a bug in the explainer when streaming queries are enabled. It is being addressed.

I can reproduce the performance difference.

**Comment by goedderz** [comment]
I actually cannot reproduce the performance difference the timing seems virtually the same. The streaming cursor seems to be slightly better even, though this may be noise.

I have executed the following in arangosh:

```
var query = `FOR d IN col LIMIT 2500000 RETURN d.i`; for (var stream of [false, true]) { var stmt = db._createStatement({query, bindVars: {}, options: {stream}}); var start = Date.now(); var cursor = stmt.execute(); while(cursor.hasNext()) { cursor.next() }; var end = Date.now(); print("stream: ", stream); print("total runtime: ", (end - start) / 1000, "s"); }
```

How did you measure the runtime?

**Comment by razbenya** [comment]
I tried your test on my data and you are right the performance issue is not reproduce via arangosh, 
tried the same thing via the java driver 

```
        String query = "for record in v_records\n" +
                "    SEARCH ANALYZER(record.queries in [\"a\"], \"text_en\")\n" +
                "    limit 2500000" +
                "    return record._key";
        for (boolean stream: Arrays.asList(false, true)) {
            long startTime = System.currentTimeMillis();
            ArangoCursor<String> cursor = db.query(query, new AqlQueryOptions().stream(stream), String.class);
            while (cursor.hasNext()) {
                cursor.next();
            }
            System.out.println("Stream: " + stream + ", Time: " + TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis() - startTime) + "s");
        }
```

```
Stream: false, Time: 13s
Stream: true, Time: 39s
```

You think the problem is in the driver ?

**Comment by goedderz** [comment]
I can't say yet. @rashtao, do you spot something here?

**Comment by Simran-B** [comment]
There is an upcoming fix: #13547 changes (for VelocyPack based protocols) the result field of cursors with `stream: true` from an unindexed to an indexed array. This is already the case for non-streaming cursors. The Java driver (as of v6.9.0) accesses those entries with linear runtime each, resulting in degraded performance as reported here.

**Comment by goedderz** [comment]
@razbenya as a workaround, you could try to configure the driver to use the protocol `Protocol.HTTP_JSON`.

**Comment by rashtao** [comment]
I have also fixed it in the Java driver replacing the direct access to vpack array items with an iterator: https://github.com/arangodb/arangodb-java-driver/commit/5c8324aa5d62f5e468e8b0e2dd0fe8c60e243172

**Comment by razbenya** [comment]
Thanks all for the support

**Comment by Simran-B** [comment]
3.7.9 adds a workaround, which will not be available in 3.8.0. There is an upcoming fix in the Java driver (6.9.0 is affected, waiting for a new release).

**Comment by rashtao** [comment]
I released a new version of the Java driver containing the fix: https://github.com/arangodb/arangodb-java-driver/releases/tag/v6.9.1

**Comment by Simran-B** [comment]
> > also the documentation recommended not using the stream options for long queries.
> 
> That is an outdated comment and was targeted at the MMFiles storage engine, which is no longer supported. Unless you're using exclusive locks or so, you can safely ignore it and use streaming queries also for long running queries.

Addressed by #13497 #13783 #13784 #13788.

With the updated driver now available, I consider this as done.
